[PM-2944] Enable Nullable For Secrets Manager

justindbaur opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

Enable [nullable reference types feature](https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references) on core Secrets Manager entities. This is a foundational step towards enabling it project wide and an ensuring we have accurate request/response models.

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

Thomas-Avery commented on src/Core/SecretsManager/Entities/ApiKey.cs:
```diff
@@ -2,22 +2,24 @@
     public Guid? ServiceAccountId { get; set; }
     [MaxLength(200)]
-    public string Name { get; set; }
+    public string Name { get; set; } = null!;
```
It seems weird initializing a non-nullable value type as null. 

Have we thought about using the `required` modifier or initializing as an empty string?

justindbaur replied:
It's a pretty common scenario to tell the compiler you know something that it doesn't, in this case that this type is often created through reflection and from data that is guaranteed in another way to be non-null. 

I'm definitely all for using `required` or even constructor parameter which both Dapper and EF will support but unfortunately that caused build errors (mostly in test code) that I felt was out of scope for now. 

I also prefer `null!` to `""` for two reasons. One, the JIT compiler knows to no-op it so it should be faster and two, because it could cover up a problem even more so. You could try and create a `new ApiKey()` and you were supposed to set the name but missed it. Now you might try to do `apiKey.Name.Trim()` and it would work with you not knowing as early that you missed assigning it an actual value. Or since it's not-null it would save into the database correctly. If it was left as null then it would throw an error.

Thomas-Avery replied:
Thank you for the detailed explanation!

I prefer `required` but I'm good with moving forward this way to get to enabling this feature project wide.

Thomas-Avery reviewed (changes requested):
It looks like this generates EF migrations to alter the columns to `nullable: false`.

Should these EF migrations be included in this PR?

justindbaur commented:
@Thomas-Avery 

> It looks like this generates EF migrations to alter the columns to nullable: false.
> 
> Should these EF migrations be included in this PR?

My plan is to try and save up the approval on all these PR's, merge them in near unison, and add all the migrations in a single PR. That way I can have only single team PR's.

Thomas-Avery reviewed (approved):
LGTM

justindbaur commented:
@coltonhurst I wasn't going to initially and it could be a team thing to improve it but since you all are all using the integration tests this change started breaking the `NOT NULL` constraints in the database for the tests that weren't setting it to a value. I went ahead and switched these to use `required` and updated the tests.

Thomas-Avery commented on test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs:
```diff
@@ -943,6 +929,22 @@ private async Task<(Guid OrganizationId, List<Guid> ServiceAccountIds)> SetupLis
+        {
+            ServiceAccountId = serviceAccountId,
+            Name = name,
+            ExpireAt = expiresAt ?? DateTime.MinValue,
```
```suggestion
            ExpireAt = expiresAt,
```

Setting `ExpireAt` to null is valid and is how we set non expiring API keys. 

Having `DateTime.MinValue` would be creating the ApiKey where it is already expired, which I don't think we would want that as the default for tests.

justindbaur replied:
Didn't even look to see that it was nullable. Good catch, fixed!

Thomas-Avery reviewed (changes requested):
Looks good, thank you!

One minor suggestion and should be good to go for me.

mzieniukbw commented on test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs:
```diff
@@ -943,6 +929,22 @@ private async Task<(Guid OrganizationId, List<Guid> ServiceAccountIds)> SetupLis
         return (org.Id, serviceAccountIds);
     }
 
+    private static ApiKey CreateTestApiKey(Guid serviceAccountId, string name = "Test Name", DateTime? expiresAt = null)
```
```suggestion
    private static ApiKey CreateTestApiKey(Guid serviceAccountId, string name = _mockEncryptedString, DateTime? expiresAt = null)
```
Name field is always encoded, this is better default

mzieniukbw commented on test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs:
```diff
@@ -943,6 +929,22 @@ private async Task<(Guid OrganizationId, List<Guid> ServiceAccountIds)> SetupLis
+            {
+                "api.secrets",
+            }.ToJsonString(),
+            EncryptedPayload = "Test Payload",
```
```suggestion
            EncryptedPayload = _mockEncryptedString,
```

mzieniukbw commented on test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs:
```diff
@@ -943,6 +929,22 @@ private async Task<(Guid OrganizationId, List<Guid> ServiceAccountIds)> SetupLis
+                "api.secrets",
+            }.ToJsonString(),
+            EncryptedPayload = "Test Payload",
+            Key = "Test Key",
```
```suggestion
            Key = _mockEncryptedString,
```

mzieniukbw commented on src/Core/SecretsManager/Entities/ApiKey.cs:
```diff
@@ -2,22 +2,24 @@
 using Bit.Core.Entities;
 using Bit.Core.Utilities;
 
+#nullable enable
```
⛏️ we usually put it at the top, before `using`
To be honest i don't know why 🤷 , do we want to be more consistent on this one ?
